src/Symfony/Component/Mailer/Bridge/Sendgrid/Webhook/SendgridRequestParser.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\Sendgrid\Webhook;
+
+use EllipticCurve\Ecdsa;
+use EllipticCurve\PublicKey;
+use EllipticCurve\Signature;
+use Symfony\Component\HttpFoundation\ChainRequestMatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcher\IsJsonRequestMatcher;
+use Symfony\Component\HttpFoundation\RequestMatcher\MethodRequestMatcher;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+use Symfony\Component\Mailer\Bridge\Sendgrid\RemoteEvent\SendgridPayloadConverter;
+use Symfony\Component\RemoteEvent\Event\Mailer\AbstractMailerEvent;
+use Symfony\Component\RemoteEvent\Exception\ParseException;
+use Symfony\Component\Webhook\Client\AbstractRequestParser;
+use Symfony\Component\Webhook\Exception\RejectWebhookException;
+
+/**
+ * @author WoutervanderLoop.nl <info@woutervanderloop.nl>
+ */
+final class SendgridRequestParser extends AbstractRequestParser
+{
+    public function __construct(
+        private readonly SendgridPayloadConverter $converter,
+        private readonly bool $validateSignature,
+    ) {
+    }
+
+    protected function getRequestMatcher(): RequestMatcherInterface
+    {
+        return new ChainRequestMatcher([
+            new MethodRequestMatcher('POST'),
+            new IsJsonRequestMatcher(),
+        ]);
+    }
+
+    protected function doParse(Request $request, string $secret): ?AbstractMailerEvent
+    {
+        $content = $request->toArray();
+        if (
+            !isset($content[0]['email'])
+            || !isset($content[0]['timestamp'])
+            || !isset($content[0]['event'])
+            || !isset($content[0]['sg_message_id'])
+        ) {
+            throw new RejectWebhookException(406, 'Payload is malformed.');
+        }
+
+        if ($this->validateSignature) {
+            if (!$request->headers->get('X-Twilio-Email-Event-Webhook-Signature')
+                || !$request->headers->get('X-Twilio-Email-Event-Webhook-Timestamp')
+            ) {
+                throw new RejectWebhookException(406, 'Signature is required.');
+            }
+
+            if (!class_exists(Ecdsa::class)) {
```

Comment:
this library also requires `ext-gmp` to work, which is neither a composer requirement nor a runtime check for now.

Reply:
it is a composer requirement of the [starkbank/ecdsa-php](https://github.com/starkbank/ecdsa-php/blob/master/composer.json) library